plugins/link/dialogs/anchor.js

```diff
@@ -27,15 +27,15 @@ CKEDITOR.dialog.add( 'anchor', function( editor ) {
 			return null;
 		}
 
-		if ( element.is( 'a' ) ) {
+		if ( element && element.data( 'cke-real-element-type' ) === 'anchor' ) {
 			return element;
 		}
 
 		// In case of table cell selection, we want to shrink selection from td to a element.
 		range.shrink( CKEDITOR.SHRINK_ELEMENT );
 		element = range.getEnclosedNode();
 
-		return element.type === CKEDITOR.NODE_ELEMENT && element.is( 'a' ) && element;
+		return element.type === CKEDITOR.NODE_ELEMENT && element && element.data( 'cke-real-element-type' ) === 'anchor';
 	}
```

Comment:
As this check is actually repeated here, let's move it into private helper, which will:

* check if `element` is set (in this case the `if` from the beginning of the function could be removed);
* check if `element` is an element;
* check if `element` is the `a` element or has appropriate value of `cke-real-element-type`.

This fix also not work in case of anchors in table selection. `getSelectedAnchor` must return an element and after your change it returns `Boolean` for anchors inside table cells. However moving `element` to the end of the `if` is also not an option (as getting `element.type` could fail due to `range.getEnclosedNode` returning `null`). The only viable solution is IMO to create mentioned above function and create normal `if` here, not that abomination (sadly it was created by me…).

I also wonder if the function could be simplified to:
```javascript
function getSelectedAnchor( selection ) {
		var range = selection.getRanges()[ 0 ],
			element = selection.getSelectedElement();

		// In case of table cell selection, we want to shrink selection from td to a element.
		range.shrink( CKEDITOR.SHRINK_ELEMENT );
		element = range.getEnclosedNode();

		if ( element && element.type === CKEDITOR.NODE_ELEMENT &&
			( element.data( 'cke-real-element-type' ) === 'anchor' ) ) {
			return element;
		}
	}
```

but I'm not sure if it won't fetch more elements than it should (e.g. getting anchor even when whole paragraph is selected).